### ISSUE TITLE
Make Advice consume IMethodAdviceContext instead of MethoadAdviceContext

### ISSUE THREAD

**Issue by shirbr510** [ArxOne/MrAdvice#63]
I have an attribute that consumes `IMethodAdvise` which I want to test.
I've tried testing it's advise method and stepped into a problem:
1. `MethodAdviceContext` has not parallel interface to mock by.
2. some of `MethodAdviceContext's` Constructor parameters are Internal (and also have not interface) and therefore I cannot instantiate them (and therefore, no testing for me :(  )

**Comment by picrap** [comment]
That's a slippy path you're following...
Couldn't you just extract the things you want to test outside the advice (or inside, but making them independant from `IMethodAdviceContext`)?

**Comment by shirbr510** [comment]
> Couldn't you just extract the things you want to test outside the advice

I should not be dependent on your implementation of `MethodAdviceContext` (remember that `IMethodAdviceContext` doesn't exist and I've offered it as a suggestion)

**for an example:**
the `Proceed` method is responsible for executing the method that is being adviced.
if I want to assert that the `Proceed` method is being called and not interrupted, I should be able to Assert that `context.Proceed()` is called once/atleast once.

any of the "metadata" of the method should also be mockable.

**for an example:**
My function's behavior changes according to the number/ type of arguments that are being served to the method.
if I can't control the amount of arguments that are being passed (without actually referring to the method itself and make methods as test cases) how can I test my own functionality?

I can give you many more examples but all I have asked for is:
either making your internal classes public, which I not advice (**pun intended**)
add simple abstractions as `IMethodAdviceContext` etc to make the work with MrAdvice easier for testing.

**Comment by picrap** [comment]
OK, even if I still do not agree, I am a kind guy and will help you :blush:.
Is it enough for you if contructors are `protected internal` instead of being `internal`-only?

**Comment by shirbr510** [comment]
You know what?
let's make it simpler for both of us and show you what I'm attempting and you'll see what's best for you:

my code looks like this:

```
public void Advise(MethodAdviceContext context)
        {
            if (_log.IsDebugEnabled)
            {
                LogMethod(context);
            }
            else
            {
                context.Proceed();
            }
        }

        public static void LogMethod(MethodAdviceContext context)
        {
            OnEnter(context);
            try
            {
                context.Proceed();
            }
            catch (Exception ex)
            {
                OnException(context, ex);
                throw;
            }
            OnExit(context);
        }

#region On Enter/Exit/Exception Logs

        private static void OnEnter(MethodAdviceContext context)
        {
            if (!context.Parameters.Any())
            {
                _log.Debug($"Entered method {context.TargetMethod.Name}");
            }
            else
            {
                _log.Debug($"Entered method {context.TargetMethod.Name} with parameters {GetParametersAsString(context)}");
            }
        }

        private static void OnExit(MethodAdviceContext context)
        {
            if (!context.HasReturnValue)
            {
                _log.Debug($"Exited method {context.TargetMethod.Name}");
            }
            else
            {
                _log.Debug($"Exited method {context.TargetMethod.Name} and returned {context.ReturnValue}");
            }
        }

        private static void OnException(MethodAdviceContext context, Exception exception)
        {
            if (string.IsNullOrWhiteSpace(exception.Message))
            {
                _log.Debug($"Method {context.TargetMethod.Name} threw an exception of type {exception.GetType()}");
            }
            else
            {
                _log.Debug(
                    $"Method {context.TargetMethod.Name} threw an exception of type {exception.GetType()} with the message \"{exception.Message}\"");
            }
        }

        #endregion

```

as I've said: I don't want to be dependant on the actual context.

the `_log` is mockable (as it is an interface).

if you make the ctors `protected internal`, it won't help me much as `ActionValues` (for an example) will still be internal and then I'll be stuck again back at square one.

**Comment by picrap** [comment]
From the methods above, which ones do you intend to test? If you test only the `OnEnter`, `OnExit` and `OnException` you can simply pass the required `MethodAdviceContext` properties instead of passing the whole instance. Same thing (a little bit heavier though) if you test the `LogMethod`, no?

**Comment by shirbr510** [comment]
True. I can pass down parameters to anything rather Easely except one case:
`context.Proceed()`

wrapping it is possible but it will look rather close to Microsoft's
`HttpContextWrapper` solution for the sealed `HttpContext` (if you know
what I'm speaking of), which I personally prefer to avoid

**Comment by picrap** [comment]
Dude, with `HttpContextWrapper`, you're getting very close to a Godwin point :laughing: 
Since I don't want to change contracts right now, I suggest everything in `*Context` classes goes `protected internal virtual` (or something close). However I'm not sure you need access to `AdviceValues` with this (and regarding your sample).

**Comment by picrap** [comment]
... I'm so screwed with those `AdviceValues`... They're everywhere... I'll think about something and let you know.

**Comment by shirbr510** [comment]
well, I'll at least need an empty constructor so I could mock proceed (or
inherit from `MethodAdviceContext` and make a stub)

**Comment by picrap** [comment]
And what if I create new valid `protected` ctors for the classes and a `virtual` `Proceed()` method? Can you work with this?

**Comment by picrap** [comment]
I did what I specified above in version 1.3.

**Comment by shirbr510** [comment]
Awesome! thank you very much :)
in the future, I hope that a full abstraction (as in interface) will
be available :)

**Comment by picrap** [comment]
Not sure, because it breaks the API.

**Comment by picrap** [comment]
However, all members could be made `virtual`.

**Comment by shirbr510** [comment]
So, I've tried to implement it with the ctor `MethodAdviceContext(IMethodAdvice methodAdvice, MethodBase targetMethod, object target, System.Type targetType, object[] parameters, AdviceContext nextAdviceContext)` and ran into an issue.

asmy parameters are:
methodAdvice = my implementation of `IMethodAdvice` (the test subject)
targetMethod = some random method
target = `new object()`
targetType = `typeof(object)`
parameters = `new object[] { 1, 2, 3, 4 }`
nextAdviceContext = a mock of `AdviceContext`

everything should work fine from this point, but suddenly, an `UnSupportedException` is thrown at initialization from the last parameter.

I've tried mocking your `AdviceContext` and failed due to:

> You cannot mock a type that does not have a public constructor because Moq will not be able to instantiate an object of that type

( source: http://stackoverflow.com/questions/3279412/mocking-a-type-with-an-internal-constructor-using-moq)

and if I'm trying to inherit from `AdviceContext` in order to create a stub, I'm facing an internal abstract method that cannot be implemented, name `Invoke`


I'm really sorry I'm giving you a hard time and I truly appreciate your treatment to the situation!
